rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,334 @@
+  seedPeers           []Multiaddr // list of seed peers to connect to initially (seeds are automatically trustlisted)
+  maxConnections      int         // maximum number of connections allowed before the connection manager begins trimming open connections
+  validationQueueSize int         // size of the queue of active pending validation messages
+  // TODO: peerExchange bool vs minaPeerExchange bool (seems like at least one of these should be deprecated)
+  //   - peerExchange == enable libp2p's concept of peer exchange in the pubsub options
+  //   - minaPeerExchange == write random peers to connecting peers
+}
+
+// == The `Stats` struct is sent on an interval via the `stats` message.
+// == It contains metrics and statistics relevant to the helper process,
+// == to be further exposed by the daemon as prometheus metrics.
+
+type Stats struct {
+  storedPeerCount       int
+  connectedPeerCount    int
+  messageMinSize        float64
+  messageMaxSize        float64
+  messageAvgSize        float64
+  latencyMin            float64
+  latencyMax            float64
+  latencyAvg            float64
+  totalBandwidthIn      float64
+  totalBandwidthOut     float64
+  totalBandwidthRateIn  float64
+  totalBandwidthRateOut float64
+}
+
+// == A `ValidationStatus` notifies the helper process of whether or not
+// == a message was valid (or relevant).
+
+type ValidationStatus int
+const (
+  VALIDATION_ACCEPT ValidationStatus = iota
+  VALIDATION_REJECT
+  VALIDATION_IGNORE
+)
+
+// == These types define the concept of an `AbstractPeerGraph`, which
+// == describes a peer traversal algorithm for the helper to perform
+// == as when finding a successful response to an RPC query.
+
+// Each node of the graph either allows any peer to query, or it
+// identifies a specific node to query..
+type AbstractPeerType int
+const (
+  ANY_PEER      AbstractPeerType = iota
+  SPECIFIC_PEER
+)
+
+// This is essentially an ADT, but we cannot encode an ADT directly
+// in Go or Cap'N Proto. An equivalent ADT definition would be:
+//   type abstract_peer_node =
+//     | AnyPeer
+//     | SpecificPeer of Peer
+type AbstractPeerNode struct {
+  typ  AbstractPeerType
+  peer Peer             // nil unless type == SPECIFIC_PEER
+}
+
+type AbstractPeerEdge struct {
+  src int
+  dst int
+}
+
+// Alternative (safer) representations are possible, but this is
+// simplest to encode in the Cap'N Proto shared schema language.
+type AbstractPeerGraph struct {
+  nodes []AbstractPeerNode
+  edges []AbstractPeerEdge
+}
+```
+
+#### Query Control Flow
+
+In contrast to the prior implementation, the query control flow in the new protocol always follows the following pattern:
+
+1) The daemon sends 1 message to the helper to begin the query (this message may instruct the helper to 1 or more requests, with control over maximum parallelism).
```

Comment:
to begin (or send out) 1 or more requests?

Reply:
Updated for clarity.